R/parse-body.R

```diff
@@ -16,230 +12,377 @@ postBodyFilter <- function(req){
 #' ```r
-#' parser <- function() {
-#'   function(value, content_type = "ct", filename, ...) {
-#'     # do something with raw value
-#'   }
+#' parser <- () {
+#'  function(value, ...) {
+#'   # do something with raw value
+#'  }
 #' }
 #' ```
 #'
-#' It should return a named list if you want values to map to
-#' plumber endpoint function args.
-#'
 #' @examples
-#' parser_json <- function() {
-#'   function(value, content_type = "application/json", ...) {
+#' # `content-type` header is mostly used to look up charset and adjust encoding
+#' parser_dcf <- function() {
+#'   function(value, content_type = "text/x-dcf", ...) {
 #'     charset <- getCharacterSet(content_type)
 #'     value <- rawToChar(value)
 #'     Encoding(value) <- charset
-#'     jsonlite::fromJSON(value)
+#'     read.dcf(value)
 #'   }
 #' }
-#' @md
+#' add_parser("dcf", parser_dcf, fixed = "text/x-dcf")
 #' @export
-addParser <- function(name, parser, pattern = NULL) {
-  if (is.null(.globals$parsers)) {
-    .globals$parsers <- list()
-  }
-  if (!is.null(.globals$parsers$func[[name]])) {
-    stop("Already have a parser by the name of ", name)
-  }
-  if (is.null(pattern)) {
-    pattern <- paste0("application/", name)
+add_parser <- function(alias, parser, fixed = NULL, regex = NULL, shortname = NULL, verbose = TRUE) {
+
+  if (!is.null(.globals$parsers[[alias]])) {
+    if (isTRUE(verbose)) {
+      warning("Overwriting parser: ", alias)
+    }
   }
-  .globals$parsers$func[[name]] <- parser
-  .globals$parsers$pattern[[name]] <- pattern
-}
 
+  stopifnot(is.function(parser))
 
+  if (length(c(fixed, regex, shortname))) {
+
+    parsers_list <- function(...) {
+
+      parser_function <- do.call(parser, list(...))
+
+      create_list <- function(names) {
+        stats::setNames(
+          replicate(
+            length(names),
+            parser_function),
+          names
+        )
+      }
+
+        parsers <- list()
+
+        if (length(shortname) > 0) {
+          parsers[[shortname]] <- parser_function
+        }
+        if (length(fixed) > 0) {
+          parsers$fixed <- create_list(fixed)
+        }
+        if (length(regex) > 0) {
+          parsers$regex <- create_list(regex)
+        }
+
+        return(parsers)
+
+    }
+
+  } else {
+
+    parsers_list = parser
 
-#' JSON
-#' @rdname parsers
-#' @export
-parser_json <- function() {
-  function(value, content_type = NULL, ...) {
-    charset <- getCharacterSet(content_type)
-    value <- rawToChar(value)
-    Encoding(value) <- charset
-    safeFromJSON(value)
   }
-}
 
+  .globals$parsers[[alias]] <- parsers_list
+
+  invisible(list_parsers())
+}
 
+#' @export
+#' @describeIn add_parser List currently registered parsers
+list_parsers <- function() {
+  .globals$parsers
+}
 
+#' @describeIn add_parser Select from global parsers and create
+#' a formatted parsers list for programmatic use.
+#' @export
+select_parsers <- function(alias = character()) {
```

Comment:
Going to make this required arg

```suggestion
select_parsers <- function(alias) {
```